File: 4.8-bc/ts/stage-3-hexagonal/src/commerce/infra/http/handler.ts
// HTTP adapter для Commerce BC.
import type { FastifyInstance } from "fastify";

import { writeError } from "../../../shared/httputil.js";
import type { Service } from "../../app/service.js";

interface PlaceReq {
  user_id?: string;
  total_cents?: number;
}

interface PlaceResp {
  order_id: string;
  status: string;
}

const UUID_RE =
  /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/i;

export class Handler {
  constructor(private readonly svc: Service) {}

  register(app: FastifyInstance): void {
    app.post<{ Body: PlaceReq }>("/orders", async (req, reply) => {
      const body = req.body ?? {};
      if (!body.user_id || !UUID_RE.test(body.user_id)) {
        return reply.code(400).send({ error: "invalid_user_id" });
      }
      if (typeof body.total_cents !== "number") {
        return reply.code(400).send({ error: "invalid_request" });
      }
      try {
        const o = await this.svc.place(body.user_id, body.total_cents);
        const resp: PlaceResp = { order_id: o.id, status: o.status };
        return reply.code(201).send(resp);
      } catch (err) {
        req.log.error({ err }, "place order failed");
        return writeError(reply, err);
      }
    });
  }
}

File: 4.8-bc/go/stage-3-hexagonal/commerce/infra/http/handler.go
package http

import (
	stdhttp "net/http"

	"github.com/go-chi/chi/v5"
	"github.com/google/uuid"

	"github.com/genkovich/claude-course-demos/4.8-bc/go/stage-3-hexagonal/commerce/app"
	"github.com/genkovich/claude-course-demos/4.8-bc/go/stage-3-hexagonal/shared/httputil"
)

type Handler struct {
	svc *app.Service
}

func NewHandler(svc *app.Service) *Handler {
	return &Handler{svc: svc}
}

func (h *Handler) Register(r chi.Router) {
	r.Post("/orders", h.place)
}

type placeReq struct {
	UserID     string `json:"user_id"`
	TotalCents int64  `json:"total_cents"`
}

type placeResp struct {
	OrderID uuid.UUID `json:"order_id"`
	Status  string    `json:"status"`
}

func (h *Handler) place(w stdhttp.ResponseWriter, r *stdhttp.Request) {
	var req placeReq
	if err := httputil.DecodeBody(r, &req); err != nil {
		httputil.WriteJSON(w, stdhttp.StatusBadRequest, map[string]string{"error": "invalid_request"})
		return
	}
	uid, err := uuid.Parse(req.UserID)
	if err != nil {
		httputil.WriteJSON(w, stdhttp.StatusBadRequest, map[string]string{"error": "invalid_user_id"})
		return
	}
	o, err := h.svc.Place(r.Context(), uid, req.TotalCents)
	if err != nil {
		httputil.WriteError(w, err)
		return
	}
	httputil.WriteJSON(w, stdhttp.StatusCreated, placeResp{OrderID: o.ID, Status: o.Status})
}
